solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpClientUtil.java

```diff
@@ -447,14 +448,39 @@ public void process(final HttpResponse response, final HttpContext context)
     }
   }
   
-  private static class GzipDecompressingEntity extends HttpEntityWrapper {
+  protected static class GzipDecompressingEntity extends HttpEntityWrapper {
+    private boolean gzipInputStreamCreated = false;
+    private InputStream gzipInputStream = null;
+
     public GzipDecompressingEntity(final HttpEntity entity) {
       super(entity);
     }
-    
+
+    /**
+     * Return a InputStream of uncompressed data.
+     * If there is an issue with the compression of the data, a null InputStream will be returned,
+     * and the underlying compressed InputStream will be closed.
+     *
+     * The same input stream will be returned if the underlying entity is not repeatable.
+     * If the underlying entity is repeatable, then a new input stream will be created.
```

Comment:
The more I think about this, the stranger this API contract seems. The caller has now way of really knowing whether they are getting an in-progress stream that is somewhere in the middle, or a new repeatable stream set to the beginning.

Reply:
Are you talking about the HttpEntity contract or the GzipDecompressingEntity contract? Because I'm just following what HttpEntities do. If you have an InputStreamEntity, there's really no way you can replay that stuff without buffering the entire thing. While a StringEntity is very easy to repeat. 

I agree the contract is strange, it should probably just always be one-time-readable.

Reply:
For more context, a [StringEntity will return a whole new InputStream every time](https://github.com/apache/httpcomponents-core/blob/master/httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/StringEntity.java#L122) and an [InputStreamEntity will return the same one every time](https://github.com/apache/httpcomponents-core/blob/master/httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/InputStreamEntity.java#L77).